storagemarket/impl/dtutils/dtutils.go

```diff
@@ -56,11 +61,37 @@ func ProviderDataTransferSubscriber(deals EventReceiver) datatransfer.Subscriber
 			if err != nil {
 				log.Errorf("processing dt event: %w", err)
 			}
+		case datatransfer.DataReceived:
+			// TODO Should we memoize here ? This can get inefficient as this will event will keep firing for the same cid if we recieve the same cid
+			// multiple times
+			cids := channelState.ReceivedCids()
+			if len(cids) != 0 {
+				set := cid.NewSet()
```

Comment:
This strikes me as potentially ineffecient -- to reconstruct the CID set on every block? At minimum it's an n^2 operation.

Data transfer state is persisted to disk. I would strongly consider, given that you are already recording received CIDs there, that you avoid recording them here as well.

It's your call, but think you are going to be happier if you simply build an API on data transfer's manager called:

```
	RestartDataTransferChannel(ctx context.Context, chid ChannelID) error
```

Anyway, I'm fine if you wanna proceed -- your call, I just think you will find yourself getting a bit stuck down this path.

If you'd like to keep the recording in here, I'd recommend taht we not record the whole list of CIDs on the data transfer -- and just provide a "LastCID" or something that updates on every DataReceived.